### DISCUSSION TITLE
Does anyone have any clue how to get this (these) libraries compiled under platformio/vscode?

### DISCUSSION THREAD

**Discussion by billy8119** [pschatzmann/arduino-audiokit#36] [Q&A] [answered]
Was kinda hoping to learn how all this works by live-stepping through the code, which arduino IDE does not seem to permit.  Alas, in platformIO, the libraries spew infinite errors when I try to compile any of the demo apps.  Has anyone had any success here?

**Comment by pschatzmann** [comment]
I tried to do some debugging on an ESP32 but my experience was just too painful: I had it working but it was just too slow to be useful.
Some times it was helpful to build it as desktop application and debug it as such.

**Reply by billy8119** [reply]
You mean live-stepping through the code? Desktop application.. hard with all the sound IO and whatnot, which is 'zactly what I need to reverse engineer and understand ;)  I was sorta hopin' the vscode, which has smarter compilation toolchains might also compile the code somewhat more rapidly if there are only minor edits between compilations... for some reason, even on my i9 it takes over a minute to compile each upload attempt.

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/wiki/Running-an-Audio-Sketch-on-the-Desktop

**Reply by billy8119** [reply]
HA! You're so amazing it makes me laugh.  Also with envy.

**Comment by billy8119** [comment]
Also, I think the link on page https://github.com/pschatzmann/arduino-audio-tools/wiki/Running-an-Audio-Sketch-on-the-Desktop should actually be :
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-desktop/generator/CMakeLists.txt
and not what it currently is:
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples-desktop/generator/CMakeLists.txt
Which is not found on server.

**Comment by billy8119** [comment]
e@emini:~/arduino-audio-tools/examples/examples-desktop/generator/build$ cmake ..
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found Regex: POSIX regex.h  
-- Could NOT find JACK (missing: JACK_INCLUDEDIR) 
-- Could NOT find ALSA (missing: ALSA_LIBRARY ALSA_INCLUDE_DIR) 
-- Found OSS: /usr/include/x86_64-linux-gnu  
-- Looking for sys/audioio.h
-- Looking for sys/audioio.h - not found
CMake Error at build/_deps/arduino_audio_tools-src/tests/CMakeLists.txt:36 (add_subdirectory):
  add_subdirectory given source
  "/home/e/arduino-audio-tools/examples/examples-desktop/generator/build/_deps/arduino_audio_tools-src/tests/mp3-helix"
  which is not an existing directory.

**Reply by billy8119** [reply]
Apparently, adding alsa libs and Jack libs does not cure this oopsie. (the screen is full of them, this is just the first one)

**Comment by pschatzmann** [comment]
I just committed a correction. The examples should compile now.
... Taking it back - I get compile error now

**Comment by billy8119** [comment]
Oh.  Once cmake blows up the first time, forever shall it blow up in future.  I added the missing deps (jack stuff and alsa sutff), then nuked the build directory, recreated it, and make works fine now.  Now it blows up unsuccessfully trying to find portaudio.h
...
/home/e/arduino-audio-tools/examples/examples-desktop/generator/build/_deps/arduino_audio_tools-src/src/AudioLibs/PortAudioStream.h:9:10: fatal error: portaudio.h: No such file or directory
    9 | #include "portaudio.h"
      |          ^~~~~~~~~~~~~
compilation terminated.

**Reply by pschatzmann** [reply]
I think you need to install portaudio on your system

**Comment by pschatzmann** [comment]
After some cleanup I have it finally working again...

**Comment by billy8119** [comment]
Yup.  Works great.  I laugh at the awesome.  Thanks! Now I can step-trace ;)

**Comment by pschatzmann** [comment, marked as answer]
Just noticed that the initial question was about PlatformIO: You find the relevant information in the Wiki!

**Reply by tookitogo** [reply]
Hi @pschatzmann,

Well, it really isn't self-explanatory that when using this library with PlatformIO, you _don't_ edit **AudioKitSettings.h**, but rather the **platformio.ini** file. The front page of the repo simply says to edit **AudioKitSettings.h** with no mention of PlatformIO. When you then scroll down and find the link to the wiki page for using PlatformIO, it is not obvious or explicit that you need to _customize_ the sample **platformio.ini** code provided, nor how to do it. While I understand that it's not the job of your documentation to teach the minutiae of PlatformIO configuration, I think it'd be helpful to give a bit more step-by-step detail.

Here's my suggested text for the wiki, since I can't edit it myself:

1. Create a new PlatformIO project. Select the **Arduino** framework and **Espressif ESP32 Dev Module** as the board type (except for ESP32-S2 and -S3 boards).
2. Add the following lines to your **platformio.ini** file:
```
lib_deps = 
    https://github.com/pschatzmann/arduino-audiokit-hal
lib_ldf_mode = deep+
build_flags = 
    -DCORE_DEBUG_LEVEL=5 
    -DAUDIOKIT_BOARD=5 
monitor_speed = 115200
```
3. Customize the **platformio.ini** file for the board you are using. The full list of the supported boards and settings can be found in [AudioKitSettings.h](https://github.com/pschatzmann/arduino-audiokit/blob/main/src/AudioKitSettings.h)
For example, to use a LyraT v4.3 board (AUDIOKIT_BOARD type 1), edit **-DAUDIOKIT_BOARD=5** to **-DAUDIOKIT_BOARD=1**. 
You don’t need to make the changes in the **AudioKitSettings.h** file itself. 
4. Create your sketch in main.cpp.
5. Compile and upload.



Regards,
Antonio

**Reply by pschatzmann** [reply]
Well, or better just read the documentation in the Wiki! Event the Readme points you to the Wiki.
I don't know what else I could do...

**Reply by tookitogo** [reply]
Hi,

I think maybe something got misunderstood just now.

“Or better just read the wiki”? Are you sure you read and understood my reply completely? I DID read the wiki and I’m telling you the wiki documentation is incomplete. That’s precisely the reason I wrote. You may feel it’s clear enough, but that’s because you know your project closely. It’s not enough for someone who’s just looking at it for the first time. 

You don’t know what else you could do?!? Huh? I went to the effort to write a clearer, more detailed text for the wiki, so that I’m not just a useless whiner. So one idea for what you could do is to replace the incomplete wiki text of the PlatformIO page with the better one I wrote and provided for you.

Regards,
Antonio

**Reply by pschatzmann** [reply]
Sorry, my mistake. I will adapt the Wiki.

**Reply by tookitogo** [reply]
Thank you! I’m sure other first-time users will appreciate it. Feel free to ask me if you need any other help with documentation, I am happy to help. (I’m not a great programmer, so I can’t contribute as much that way, so I try to support great projects in other ways. :) ) 
Regards,
Antonio

**Comment by pschatzmann** [comment]
I adjusted the description: The whole thing was much easier when I just supported the AudioKit
